internal/graph/check.go

```diff
@@ -981,6 +988,466 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+	defer span.End()
+
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	req.GetRequestMetadata().Depth--
+	if req.GetRequestMetadata().Depth == 0 {
+		return nil, ErrResolutionDepthExceeded
+	}
+
+	commonParameters.dsCount.Add(1)
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	objectToUsersetMessageChan := streamedLookupUsersetForObject(
+		cancellableCtx,
+		commonParameters.typesys,
+		commonParameters.ds,
+		req,
+		commonParameters.allowedUserTypeRestrictions,
+		commonParameters.concurrencyLimit)
+
+	var usersetItems []string
+	for usersetMsg := range objectToUsersetMessageChan {
+		if usersetMsg.err != nil {
+			return nil, usersetMsg.err
+		}
+		usersetName := usersetMsg.userset
+		if commonParameters.userToUsersetMapping.Exists(usersetName) {
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolveCheckResponseMetadata{
+					DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
+				},
+			}, nil
+		}
+		usersetItems = append(usersetItems, usersetName)
+	}
+
+	if len(usersetItems) == 0 {
+		return &ResolveCheckResponse{
+			Allowed: false,
+			ResolutionMetadata: &ResolveCheckResponseMetadata{
+				DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + commonParameters.dsCount.Load(),
+			},
+		}, nil
+	}
+
+	return parallelizeRecursiveMatchUserUserset(ctx, usersetItems, req, commonParameters, recursiveMatchUserUserset)
+}
+
+type usersetMessage struct {
+	userset string
+	err     error
+}
+
+// streamedLookupUsersetForUser streams the userset (req's object#relation) that are assigned to
+// the user to the usersetMessageChan channel.
+func streamedLookupUsersetForUser(ctx context.Context,
+	typesys *typesystem.TypeSystem,
+	ds storage.RelationshipTupleReader,
+	req *ResolveCheckRequest,
+	maxConcurrentReads int,
+) chan usersetMessage {
+	ctx, span := tracer.Start(ctx, "streamedLookupUsersetForUser")
+	defer span.End()
+
+	usersetMessageChan := make(chan usersetMessage, maxConcurrentReads)
+
+	go func() {
+		defer func() {
+			close(usersetMessageChan)
+		}()
+
+		// Note that if the type of the request's user is publicly assignable, this will fetch that tuple as well.
+		iter, err := checkutil.IteratorReadStartingFromUser(ctx,
+			typesys, ds, req,
+			tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
+			nil)
+		if err != nil {
+			span.RecordError(err)
+			concurrency.TrySendThroughChannel(ctx, usersetMessage{
+				userset: "",
+				err:     err,
+			}, usersetMessageChan)
+			return
+		}
+
+		filteredIter := storage.NewConditionsFilteredTupleKeyIterator(
+			storage.NewFilteredTupleKeyIterator(
+				storage.NewTupleKeyIteratorFromTupleIterator(iter),
+				validation.FilterInvalidTuples(typesys),
+			),
+			checkutil.BuildTupleKeyConditionFilter(ctx, req.GetContext(), typesys),
+		)
+		defer filteredIter.Stop()
+
+		for {
+			if ctx.Err() != nil {
```

Comment:
Why this check here?

Reply:
In `streamedLookupUsersetForObject` you rely on `.Next` telling you its cancelled...

Reply:
Right, rely on next to tell it is cancelled.